Restructuring `topic_links` format returned by our API.

sumanthvrao opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
The linkifier code now includes both the shortened text and the expanded URL, sorted by the order of the occurrence in a topic. This list is passed back in the `topic_links` parameter of the /messages and the /events APIs. 

```
earlier: ['https://www.google.com', 'https://github.com/zulip/zulip/32']

now: [{'url': 'https://www.google.com", 'text': 'https://www.google/com},
         {'url': 'https://github.com/zulip/zulip/32', 'text': '#32'}]
```

**Testing plan:** <!-- How have you tested? -->
Tests amended and verified manually.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->
Fixes: #17109

aero31aero commented on static/templates/recipient_row.hbs:
```diff
@@ -32,7 +32,7 @@
                 {{! exterior links (e.g. to a trac ticket) }}
                 {{#each topic_links}}
-                <a href="{{this}}" target="_blank" rel="noopener noreferrer" class="no-underline">
+                <a href="{{this.url}}" target="_blank" rel="noopener noreferrer" class="no-underline">
```
minor: this change is in the commit that's advertised as just the API change.

aero31aero commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2083,15 +2083,19 @@ def make_md_engine(realm_filters_key: int, email_gateway: bool) -> None:
         pattern = prepare_realm_pattern(realm_filter[0])
         for m in re.finditer(pattern, topic_name):
-            matches += [realm_filter[1] % m.groupdict()]
+            matched_details = m.groupdict()
```
minor: `match_details` sounds more natural to me, even more so because we're already iterating over a match object (`for m in re.finditer()`).

aero31aero commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2083,15 +2083,19 @@ def make_md_engine(realm_filters_key: int, email_gateway: bool) -> None:
-            matches += [realm_filter[1] % m.groupdict()]
+            matched_details = m.groupdict()
+            matched_text = matched_details["linkifier_actual_match"]
+            matches += [dict(url=realm_filter[1] % matched_details,
```
I've previously thought this construction isn't very easy to read. In fact, even when reading this code yesterday, I glanced over it as code that somehow works but I'm not interested in looking up the specifics of how the regex match API works, while _all it is is a format string_.

I'm not sure how others feel about it but I feel we should have a follow-up commit to explicitly say what we're doing, like:

```python
# format the realm_filter's url string using the matched text.
url = realm_filter[1] % match_details
```

sumanthvrao replied:
I was also confused as to what was going on here the first time I read it. I added a similar comment and one below as to why we are including the match_text. Thanks for pointing it out :)

aero31aero reviewed (approved):
This looks great @sumanthvrao and exactly like what I was thinking. After reviewing all three commits, I've left a couple of small comments.

sumanthvrao commented:
@aero31aero Thanks for the review! Addressed your feedback with an update I pushed back here.

aero31aero commented on zerver/lib/markdown/__init__.py:
```diff
@@ -2083,15 +2083,22 @@ def make_md_engine(realm_filters_key: int, email_gateway: bool) -> None:
 # are validated and escaped inside `url_to_a`).
-def topic_links(realm_filters_key: int, topic_name: str) -> List[str]:
-    matches: List[str] = []
+def topic_links(realm_filters_key: int, topic_name: str) -> List[Dict[str, Union[str, int]]]:
```
This type doesn't need a union; we're stripping away the `index` field before returning.

sumanthvrao replied:
Indeed, however, mypy shouts at me with 
```
(zulip-py3-venv) vagrant@23d7671b6078:/srv/zulip$ ./tools/run-mypy
zerver/lib/markdown/__init__.py:2121: error: Incompatible return value type 
(got "List[Dict[str, Union[str, int]]]", expected "List[Dict[str, str]]")  [return-value]
Found 1 error in 1 file (checked 1184 source files)
```
Wondering if it Is because I am returning back the same `matches` object? Any ideas?

aero31aero replied:
Yeah, that makes sense. Maybe copying to a new List would help here. I'll run some experiments. Maybe this is something they've addressed already but it does look like a hard problem to track which elements of each dict have been deleted with just static type checks so I doubt there's a solution without cloning the List.

timabbott replied:
Yeah, mypy doesn't like changing the type of a variable.  Generally code is more readable using a new variable name for the output (vs. the input), anyway, so I'd recommend doing that.

sumanthvrao replied:
Hmmm... this is proving to be tricky. Seems like even using a new variable and copying to it the k, v pair seems to unsettle mypy. 
```
    .....
    matches = sorted(matches, key=lambda k: k["index"])
    matches_output: List[Dict[str, str]] = [{k: v for k, v in match.items() if k != "index"} for match in matches]
    return matches_output
```
still throws 
```
zerver/lib/markdown/__init__.py:2118: error: Value expression in dictionary comprehension has incompatible type "Union[str, int]"; expected type "str"  [misc]
Found 1 error in 1 file (checked 1184 source files)
```

Hmmm...Seems like only type-casting the value in the dictionary before returning seems to work.
```
    return [{k: str(v) for k, v in match.items() if k != "index"} for match in matches]
```

timabbott replied:
Hmm, yeah, I think that's reasonable.

timabbott commented on templates/zerver/api/changelog.md:
```diff
@@ -15,6 +15,10 @@ below features are supported.
+* [`GET /messages`](/api/get-messages) and [`GET
+  /events`](/api/get-events): `topic_links` returns a list
+  of objects each containing the `url` and `text` for links
+  that are present in the topic or generated by linkifiers.
```
This needs a corresponding change to the documentation in `zerver/openapi/zulip.yaml`.

sumanthvrao commented:
@timabbott the changes you mentioned have been added, so this is now pending a review by you.

timabbott commented:
This is great!  A few requests:
* Can you squash the commits?  It's structurally a signal atomic change -- just the different files that need updating for a single feature.
* Can you update ZULIP_FEATURE_LEVEL to 40 with this?  Basically a rebase consequence.  I'll also add another comment on the Changes sections.
* Can you make sure we update the `zulip-mobile` issue for using `topic_links` to mention this change?
* Did we want to discuss a second commit to e.g. make use of the `title` attribute to indicate what generated the links in the 
Zulip web client?  I can't remember the discussion on that topic.

timabbott commented on templates/zerver/api/changelog.md:
```diff
@@ -15,6 +15,10 @@ below features are supported.
   to 259200s (3 days).
+* [`GET /messages`](/api/get-messages) and [`GET
+  /events`](/api/get-events): `topic_links` returns a list
+  of objects each containing the `url` and `text` for links
```
Maybe ": `topic_links` fields now contain a list of objects, rather than a list of strings".?  think the details of the change are best expressed on the page themselves; this is meant to be a very high level reference.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -1859,11 +1859,16 @@ paths:
+                                    **Changes**: New in Zulip 4.0 (feature level 38).
+                                      Previously, this field returned a list of urls. Now,
+                                      it contains a list of dictionaries containing the matched
+                                      url and text for each link in the `topic`.
```
The phase "New in Zulip ..." is meant to be used only when the field is new.  I think we can replace this line with:

"**Changes**: This field contained a list of URLs before Zulip 4.0 (feature level 40)."  We don't need to repeat the details of the new format, because that's explained in the main text of the documentation; we just need to explain what the old format meant.

And then I like your approach of keeping the old Changes as a new paragraph.

timabbott replied:
I think you ended up collapsing the old changes into the existing paragraph; I undid that (see the screenshot in the issue I opened below).

sumanthvrao commented:
@timabbott updated the PR with some additional changes:
* I added to our a couple of new tests that focus on testing the order in which the links inside a topic occur.
* Also, we had missed out on changing the openAPI schema to match the new object data-type returned. I have added this as well.

timabbott commented:
This is great, merged, thanks @sumanthvrao!  I did a handful of things before merging:

* Resolved the merge conflict and made this feature level 46, updating `version.py` too.
* Updated the docs to say feature level 46, not 40, since since you missed that in the 40=>45 rebase.  :)
* I opened https://github.com/zulip/zulip/issues/17895 for a visual issue with our API docs that I noticed while reviewing this
* I did the following edit for better documentation English (adding "the").
```
-  /events`](/api/get-events): `topic_links` field now contains a list
-  of objects, rather than a list of string.
+  /events`](/api/get-events): The `topic_links` field now contains a
+  list of dictionaries, rather than a list of strings.
```
* I did a bit of variable name cleanup for better readability:
```
diff --git a/zerver/lib/markdown/__init__.py b/zerver/lib/markdown/__init__.py
index d907092faf..c3009c0e7f 100644
--- a/zerver/lib/markdown/__init__.py
+++ b/zerver/lib/markdown/__init__.py
@@ -2387,7 +2387,9 @@ def topic_links(realm_filters_key: int, topic_name: str) -> List[Dict[str, str]]
     realm_filters = realm_filters_for_realm(realm_filters_key)
 
     for realm_filter in realm_filters:
-        pattern = prepare_realm_pattern(realm_filter[0])
+        raw_pattern = realm_filter[0]
+        url_format_string = realm_filter[1]
+        pattern = prepare_realm_pattern(raw_pattern)
         for m in re.finditer(pattern, topic_name):
             match_details = m.groupdict()
             match_text = match_details["linkifier_actual_match"]
@@ -2396,7 +2398,7 @@ def topic_links(realm_filters_key: int, topic_name: str) -> List[Dict[str, str]]
             # don't have to implement any logic of their own to get back the text.
             matches += [
                 dict(
-                    url=realm_filter[1] % match_details,
+                    url=url_format_string % match_details,
                     text=match_text,
                     index=topic_name.find(match_text),
                 )
@@ -2406,13 +2408,15 @@ def topic_links(realm_filters_key: int, topic_name: str) -> List[Dict[str, str]]
     for sub_string in basic_link_splitter.split(topic_name):
         link_match = re.match(get_web_link_regex(), sub_string)
         if link_match:
-            url = actual_match_url = link_match.group("url")
-            result = urlsplit(url)
+            actual_match_url = link_match.group("url")
+            result = urlsplit(actual_match_url)
             if not result.scheme:
                 if not result.netloc:
                     i = (result.path + "/").index("/")
                     result = result._replace(netloc=result.path[:i], path=result.path[i:])
                 url = result._replace(scheme="https").geturl()
+            else:
+                url = actual_match_url
             matches.append(
                 dict(url=url, text=actual_match_url, index=topic_name.find(actual_match_url))
             )
```

sumanthvrao commented:
Thanks for the tweaks! Yeah I had done a half-rebase previously and was yet to get to fixing up the CI error (Thanks for pointing that out). I also updated the `zulip-terminal` issue which planned to implement this (I think a PR is already out there) with the new topic links structure.
